Re-add initial Flash Attention support on ROCM

xinyazhang opened the pull request:
Note about the Updates:

This PR:
1. skips more flash attention related UTs on MI200
2. Fix additional ATen compiling errors after hipification
3. Fix the author "root" of a specific commit
4. Includes the patch from Nikita in favor of block level static initialization.

CAVEAT: This revised PR has a commit that modifies the CI to force its running on MI200 nodes. That specific commit must be reverted before merge.

Original PR (https://github.com/pytorch/pytorch/pull/114309) Note:

This pull requests add initial Flash Attention support for AMD/ROCM platform. It added a specialized Triton repository/branch as a compile-time dependency for Flash Attention math library on AMD/ROCM. This triton submodule is not used at runtime and will not be shipped to the final pytorch package. We have the plan to release this specialized Triton as a separate project.

Know limitations:

- Only supports MI200 series GPU (i.e., `gcnArchName == gfx90a:sramecc+:xnack-`.
- Only supports power of two sequence lengths.
- No support for varlen APIs.
- Only support head dimension 16,32,64,128.
- Performance is still being optimized.

Fixes #112997

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

groenenboomj committed:
Experimental Flash Attention support infrastructure for ROCM

xinyazhang committed:
Fixes and code clean up for the initial support infrastructure.

xinyazhang committed:
Selective hipification of transformers

xinyazhang committed:
test_transformers: separate flash attention from memory efficient attention.

groenenboomj committed:
Add flash attention support for ROCM.

Also clean up the code and unused files.

xinyazhang committed:
Integrate flash attention on ROCM as a third_part library.

xinyazhang committed:
Code clean up.

xinyazhang committed:
Various fixes and clean ups

1. Skip unsupported flash attention tests on ROCM.
2. Fix the invocation of flash attention backend.
3. Change flash attention backend (oort) into a cmake External project from
   a submodule under third_party/
4. Fix error: reinterpret_cast from type const c10::complex<double>* to
   type hipblasDoubleComplex* casts away qualifiers
5. Guard Triton Flash Attention with USE_FLASH_ATTENTION in cmake

groenenboomj committed:
Pass a subset of flash attention tests

malfet committed:
[BE][ROCm] Use modern C++ (#115844)

This removes `over_arch` and `gcn_arch_override_flag` variables in favor of block level static initialization introduced in C++11

To quote from [ISO/IEC 14882-2014](https://www.open-std.org/jtc1/sc22/wg21/docs/standards)
>The zero-initialization (8.5) of all block-scope variables with static storage duration (3.7.1) or thread storage
> duration (3.7.2) is performed before any other initialization takes place. Constant initialization (3.6.2) of a
> block-scope entity with static storage duration, if applicable, is performed before its block is first entered.
> An implementation is permitted to perform early initialization of other block-scope variables with static or
> thread storage duration under the same conditions that an implementation is permitted to statically initialize
> a variable with static or thread storage duration in namespace scope (3.6.2). Otherwise such a variable is
> initialized the first time control passes through its declaration; such a variable is considered initialized upon
> the completion of its initialization. If the initialization exits by throwing an exception, the initialization
> is not complete, so it will be tried again the next time control enters the declaration. If control enters
> the declaration concurrently while the variable is being initialized, the concurrent execution shall wait for
> completion of the initialization.

xinyazhang committed:
!!! MUST DROP BEFORE MERGING !!! Force CI running on MI200.

xinyazhang commented:
Note to maintainer.

This PR has a commit that modifies the CI to force it running on MI200 nodes. This commit must be removed before merge.

xinyazhang commented:
@pytorchbot label ciflow/binaries_wheel ciflow/trunk "release notes: rocm"

xinyazhang commented:
Again: DO NOT merge this without dropping commit https://github.com/pytorch/pytorch/pull/115981/commits/d028c5b7b8136cd38f478cf1a458a28d2d5c961c

xinyazhang commented:
@pytorchbot label keep-going

malfet commented on caffe2/CMakeLists.txt:
```diff
@@ -955,6 +955,7 @@ endif()
 if(USE_ROCM)
   set(CUDA_LINK_LIBRARIES_KEYWORD PRIVATE)
   hip_add_library(torch_hip ${Caffe2_HIP_SRCS})
+  target_link_libraries(torch_hip PRIVATE __caffe2_oort)
```
Shouldn't this be guarded by `if(USE_FLASH_ATTENTION)`?

xinyazhang replied:
@malfet After trying for a few more cycles, no, we can't, unfortunately.

`USE_FLASH_ATTENTION` and `USE_MEM_EFF_ATTENTION` are used in `cmake/Dependencies.cmake` to enable FA/MEFF, but these two flags depends on `USE_ROCM`, which defaults to True and will be set correctly inside `cmake/Dependencies.cmake`.

malfet commented on cmake/Dependencies.cmake:
```diff
@@ -1295,6 +1295,7 @@ if(USE_ROCM)
       message(STATUS "Disabling Kernel Assert for ROCm")
     endif()
 
+    include(${CMAKE_CURRENT_LIST_DIR}/External/oort.cmake)
```
Should this be guarded by `if(USE_FLASH_ATTENTION)`

malfet commented on .github/workflows/rocm.yml:
```diff
@@ -25,12 +25,12 @@ jobs:
-          { config: "default", shard: 4, num_shards: 6, runner: "linux.rocm.gpu" },
-          { config: "default", shard: 5, num_shards: 6, runner: "linux.rocm.gpu" },
-          { config: "default", shard: 6, num_shards: 6, runner: "linux.rocm.gpu" },
+          { config: "default", shard: 1, num_shards: 6, runner: "linux.rocm.gpu.2" },
+          { config: "default", shard: 2, num_shards: 6, runner: "linux.rocm.gpu.2" },
+          { config: "default", shard: 3, num_shards: 6, runner: "linux.rocm.gpu.2" },
+          { config: "default", shard: 4, num_shards: 6, runner: "linux.rocm.gpu.2" },
+          { config: "default", shard: 5, num_shards: 6, runner: "linux.rocm.gpu.2" },
+          { config: "default", shard: 6, num_shards: 6, runner: "linux.rocm.gpu.2" },
```
From the PR description, it looks like this change should be reverted, shouldn't it?

jeffdaily replied:
It will be, yes.  IIUC we've confirmed the MI210s have given a green signal now, so we will revert this change.

malfet commented:
@pytorchbot merge

xinyazhang commented:
@malfet Wait a sec I haven't reverted the MI210 CI change.

xinyazhang committed:
cmake: Fix USE_FLASH_ATTENTION and USE_MEM_EFF_ATTENTION.

They are used as flags for user inputs, but will be disabled by
caffe2_update_option on ROCM regardless initial user settings.

xinyazhang committed:
Revert "!!! MUST DROP BEFORE MERGING !!! Force CI running on MI200."

This reverts commit 1a76dbeb5549f809d8d72355573a17e65f44f342.

jithunnair-amd commented:
All unit tests passed on MI210 in `rocm` workflow:
https://hud.pytorch.org/pytorch/pytorch/pull/115981?sha=1a76dbeb5549f809d8d72355573a17e65f44f342

jithunnair-amd commented:
@pytorchbot merge -f "Unrelated CI failures"
